c2rust-analyze/src/dataflow/type_check.rs

```diff
@@ -284,6 +284,35 @@ impl<'tcx> TypeChecker<'tcx, '_> {
 
                     Some(Callee::MiscBuiltin) => {}
 
+                    Some(Callee::Malloc) => {
+                        self.visit_place(destination, Mutability::Mut);
+                    }
+
+                    Some(Callee::Calloc) => {
+                        self.visit_place(destination, Mutability::Mut);
+                    }
+                    Some(Callee::Realloc) => {
+                        // We handle this like a pointer assignment.
```

Comment:
`realloc` potentially invalidates the incoming pointer, so I think we need to treat it as a combined `free` and `malloc`.  Since the data inside the allocation is either preserved / copied over, we also need to unify non-outermost pointer types, similar to an assignment.

For a `realloc` that takes a `*mut /*1*/ *mut /*2*/ i32` and returns a `*mut /*3*/ *mut /*4*/ i32`, `PointerId(2)` and `PointerId(4)` (the non-outermost pointer types) should be unified, `PointerId(1)` should be treated like an argument to `free` (probably requiring `FREE` permission), and `PointerId(3)` should be treated like the return value of a `malloc` (which I don't think requires any special handling).

Reply:
is something like 4ddbd096 what you had in mind?